pkg/app/services/subscription/service.go

```diff
@@ -0,0 +1,354 @@
+		if err == gorm.ErrRecordNotFound {
+			return newTrialSubInfo(int(trialPeriod.Hours() / 24)), nil
+		} else if err != nil {
+			return nil, errors.Wrap(err, "failed to fetch oldest unscoped org sub")
+		}
+
+		return newTrialSubInfo(int((trialPeriod - time.Since(sub.CreatedAt)).Hours() / 24)), nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch scoped org sub")
+	}
+
+	return &WrappedSubInfo{Subscription: returntypes.SubFromModel(sub)}, nil
+}
+
+func (s *basicService) Get(rc *request.AuthorizedContext, context *request.OrgSubID) (*returntypes.SubInfo, error) {
+	var org models.Org
+	if err := models.NewOrgQuerySet(rc.DB).IDEq(context.OrgID.OrgID).One(&org); err != nil {
+		return nil, errors.Wrap(err, "failed to get org from db")
+	}
+	if err := s.isAdminCached(rc, &org); err != nil {
+		return nil, errors.Wrap(err, "failed to check for admin")
+	}
+
+	var sub models.OrgSub
+	err := models.NewOrgSubQuerySet(rc.DB).IDEq(context.SubID.SubID).One(&sub)
+	if err == gorm.ErrRecordNotFound {
+		return nil, apierrors.ErrNotFound
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch scoped org sub")
+	}
+
+	return returntypes.SubFromModel(sub), nil
+}
+
+func (s basicService) finishQueueSending(rc *request.AuthorizedContext, sub *models.OrgSub,
+	expState, setState models.OrgSubCommitState) (*returntypes.SubInfo, error) {
+
+	n, err := models.NewOrgSubQuerySet(rc.DB).
+		IDEq(sub.ID).CommitStateEq(expState).
+		GetUpdater().
+		SetCommitState(setState).
+		UpdateNum()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to update commit state to %s for sub with id %d",
+			setState, sub.ID)
+	}
+	if n == 0 {
+		rc.Log.Infof("Not updating sub %#v commit state to %s because it's already updated by queue consumer",
+			sub, setState)
+	}
+	sub.CommitState = setState
+
+	return returntypes.SubFromModel(*sub), nil
+}
+
+func (s basicService) sendToCreateQueue(rc *request.AuthorizedContext, sub *models.OrgSub) (*returntypes.SubInfo, error) {
+	if err := s.CreateQueue.Put(sub.ID); err != nil {
+		return nil, errors.Wrap(err, "failed to put to create repos queue")
+	}
+	return s.finishQueueSending(rc, sub, models.OrgSubCommitStateCreateInit, models.OrgSubCommitStateCreateSentToQueue)
+}
+
+//nolint:gocyclo
+func (s *basicService) Create(rc *request.AuthorizedContext, context *request.OrgID, payload *SubPayload) (*returntypes.SubInfo, error) {
+	if payload.PaymentGatewayCardToken == "" || payload.IdempotencyKey == "" {
+		return nil, errors.New("idempotency key and card token are required for new subscriptions")
+	}
+
+	var org models.Org
+	if err := models.NewOrgQuerySet(rc.DB).IDEq(context.OrgID).One(&org); err != nil {
+		return nil, errors.Wrap(err, "failed to get org from db")
+	}
+	if err := s.isAdminCached(rc, &org); err != nil {
+		return nil, errors.Wrap(err, "failed to check for admin")
+	}
+
+	var retSub *returntypes.SubInfo
+
+	// TODO: This is probably not the best way to do this.
```

Comment:
It could be done easier and more reliable: add `idempotency_key` field to table `subs` with unique constraint on it. If insert failed with constraint error - we know it's the repeated request.